## **Building first Neural Network with TensorFlow and Keras**

#### **Things to know:**

#### Definition of terms

1. Sequential(): A Sequential model is appropriate for a plain stack of layers where each layer has exactly one input tensor and one output tensor.

2. Dense(): A dense layer is a layer that is deeply connected with its preceding layer which means the neurons of the layer are connected to every neuron of its preceding layer.

**Arguments:**

**units**: Positive integer, dimensionality of the output space <br>
**activation**: Activation function to use. If you don't specify anything, no activation is applied (ie. "linear" activation: a(x) = x) <br>
**use_bias**: Boolean, whether the layer uses a bias vector. <br>
**kernel_initilizer**: Initializer for the kernel weights matrix. <br>
**bias_initializer**: Initializer for bias vector <br>
**bias_regularizer**: Regularizer function applied to the bias vector. <br>
**activity_regularizer**: Regularizer function applied to the output of the layer (its "activation") <br>
**kernel_constraint**: Constraint function applied to the kernel weights matrix <br>
**bias_constraint**: Constraint function applied to the bias vector. <br>

Compile(): putting all our recipes together and making it ready for training in the next phase

*model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
              loss=tf.keras.losses.BinaryCrossentropy(),
              metrics=[tf.keras.metrics.BinaryAccuracy(),
                       tf.keras.metrics.FalseNegatives()])*

fit(): Trains the model for a fixed number of epochs (iterations on a dataset). <br>

1. **epoch**: One pass through all of the rows in the training dataset. <br>
2. **batch**: one or more samples considered by the model within an epoch before weights are updated. <br>

## **What is a Sample?**
A sample is a single row of data. <br>

It contains inputs that are fed into the algorithm and an output that is used to compare to the prediction and calculate an error. <br>

A training dataset is comprised of many rows of data, e.g. many samples. A sample may also be called an instance, an observation, an input vector, or a feature vector. <br>

Now that we know what a sample is, let's define a batch.

## **What is a Batch?**

The batch is a hyperparameter that defines the number of samples to work through before updating the internal model parameters. <br>

Thingk of a batch as a for-loop iterating over one or more samples and making predictions. At the end of the batch, the predictions are compared to the expected output variables and an error is calculated. From this error, the update algorithm is used to improve the model, e.g. move down along the error gradient. <br>

A training dataset can be divided into one or more batches. <br>

When all training samples are used to create one batch, the learning algorithm is called batch gradient descent. When the batch is the size of one sample, the learning algorithm is called stochastic gradient descent. When the batch size is more than one sample and less than the size of the training data set, the learning algorithm is called mini-batch gradient descent <br>

1. **Batch Gradient Descent**: Batch Size = Size of Training Set <br>
2. **Stochastic Gradient Descent**: Batch Size = 1 <br>
3. **Mini-Batch Gradent Descent**: Batch Size < Size of Training Set <br>

In the case of mini-bath gradient descent, popular batch sizes include 32,64, and 128 samples. You may see these values used in models in literature and in tutorials.



## **What is an Epoch?**
The number of epochs is a hyperparameter that defines the number of times that he learning algorithm will work through the entire training dataset. <br>

One epoch means that each sample in the training dataset has had an opportunity to update the internal model parameters. An epoch is comprised of one or more batches. For example, as above, an epoch that has one batch is called the batch gradient descent learning algorithm. <br>

You can think of a for-loop over the number of epochs where each loop proceeds over the training dataset. Within ths for-loop is another nested for-loop that iterates over every batch of samples, where one batch has the specified "batch size" number of samples. <br>

The number of epochs is traditionally large, often hundreds or thousands, allowing the learning algorithm to run until the error from the model has been sufficiently minimized. You may see examples of the number of epochs in the literature and in tutorials set to 10, 100, 500, 1000 and larger.



## **What is the Difference between Batch and Epoch**
1) The batch size is a number of samples processed before the model is updated <br>
2) The number of eopochs is the number of complete passes through the training dataset <br>
3) The size of a batch must be more than or equal to one or less than or equal to the number of samples in the training dataset. <br>
4) The number of epochs can be set to an integer value between one and infinity. You can run the algorithm for as long as you like and even stop it using other criteria besides a fixed number of epochs, such as change (or lack of change) in model error over time.
5) They are both integer in values and are both hyperparametersfor the learning algorithm, e.g. parameters for th learning process, not internal model parameters found by the learning process.
6) You must specify the batch size and number of epochs for a learning algorithm.
7) There are no magic rules for how to configure these parameters. You must try different values and see what works best for your problem.

## **Worked Example**

Finally, let's make this concrete with a small example. <br>

Assume you have a dataset with 200 samples (rows of data) and you choose a batch size of 5 and 1,000 epoch <br>

This means that the dataset will be divided into 40 batches, each with five samples. The model wights will be updated after each batch of five samples. <br>

This also means that one epoch will involve 40 batches or 40 updates to the model. <br>

With 1000 epochs, the model will be exposed to or pass through the whole dataset 1,000 times. That is a total of 40,000 batches during the entire training process. <br>

Note the folloring steps in creating your neural network: <br>
1. Define our training data <br>
2. Create a sequential model: <br>
3. Add layers <br>
4. Compile our model <br>
5. Predict values <br>

## **Building A Neural Network**

## **1. Load the Data**

In [ ]:
pip install tensorflow

pip install keras

In [18]:
import numpy as np
from numpy import loadtxt
import keras
from keras.models import Sequential
from keras.layers import Dense

In [23]:
dataset = loadtxt("pima-indians-diabetes.data.csv", delimiter = ',')
X = dataset [:,0:8]
y= dataset[:,8]

In [24]:
y

array([1., 0., 1., 0., 1., 0., 1., 0., 1., 1., 0., 1., 0., 1., 1., 1., 1.,
       1., 0., 1., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 1., 0., 0.,
       0., 0., 0., 1., 1., 1., 0., 0., 0., 1., 0., 1., 0., 0., 1., 0., 0.,
       0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0.,
       0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 1., 1., 1., 0., 0.,
       0., 1., 0., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0.,
       1., 1., 0., 0., 0., 1., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 1.,
       1., 1., 1., 0., 0., 1., 1., 0., 1., 0., 1., 1., 1., 0., 0., 0., 0.,
       0., 0., 1., 1., 0., 1., 0., 0., 0., 1., 1., 1., 1., 0., 1., 1., 1.,
       1., 0., 0., 0., 0.

**Note:** the dataset has 9 columns and the range 0:8 will select columns from 0 to 7, stopping before index 8.

In [26]:
X

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [28]:
import pandas as pd
X_df = pd.DataFrame(X)
X_df

0      1     2     3      4     5      6     7
0     6.0  148.0  72.0  35.0    0.0  33.6  0.627  50.0
1     1.0   85.0  66.0  29.0    0.0  26.6  0.351  31.0
2     8.0  183.0  64.0   0.0    0.0  23.3  0.672  32.0
3     1.0   89.0  66.0  23.0   94.0  28.1  0.167  21.0
4     0.0  137.0  40.0  35.0  168.0  43.1  2.288  33.0
..    ...    ...   ...   ...    ...   ...    ...   ...
763  10.0  101.0  76.0  48.0  180.0  32.9  0.171  63.0
764   2.0  122.0  70.0  27.0    0.0  36.8  0.340  27.0
765   5.0  121.0  72.0  23.0  112.0  26.2  0.245  30.0
766   1.0  126.0  60.0   0.0    0.0  30.1  0.349  47.0
767   1.0   93.0  70.0  31.0    0.0  30.4  0.315  23.0

[768 rows x 8 columns]

## **2.Define the Keras Model**
The first thing to get right is to ensure the input layer has the right number of input features. This can be specified when creating the first layer with the **input_dm** argument and setting it to 8 for the 8 input variables. <br>

How do we know the number of layers and their types? <br>

This is a very hard question. There are heuristics that we can use and often the best network structure is found through a process of trial and error experimentation. Generally, you need a network large enough to capture the structure of the problem. <br>

In this example, we will use a fully-connected network structure with three layers. <br>

Fully connected layers are defined using the Dense class. We can specify the number of neurons or nodes in the layer as the first argument, and specify the activation function using the activation argument. <br>

We will use the rectified linear unit activation function related to as ReLU on the first two layers and the Sigmoid function in the output layer. <br>

It used to be the case that Sigmoid function and Tanh activation functions were preferrred for all layers. These days, better performance is achieved using the ReLU activation function. We use a sigmoid on the outer layer to ensure out network output is between 0 and 1 and easy to map to either a probability of class 1 or snap to a hard classification of either class with a default threshold of 0.5 <br>

We can piece it all together by adding each layer: <br>
1. The model expects rows of data with 8 variables (the input_dim=8 argument) <br>
2. The first hidden layer has 12 nodes and uses the relu activation function. <br>
3. The second hidden layer has 8 nodes and uses the relu activation function. <br>
4. The output layer has one node and uses the sigmoid activation function.

In [30]:
model = Sequential()
model.add(Dense(units=12, input_dim=8, activation='relu')) #1st hidden layer
model.add(Dense(units=9, activation='relu')) #2nd hidden layer
model.add(Dense(units=1,activation='sigmoid')) #3rd hidden layer

**Note**: the most confusin thing here is that the shape of the input to the model is defined as the first argument on the first hidden layer. This means that the line of code that adds the first Dense layer is doing 2 things, defining the input or visible layer and the first hidden layer.

## **3. Compile keras Model**
Now that the model is defined, we can compile it.

We must specify the loss function to use to evaluate a set of weights, the opimizer is used to search through different weights for the network and any optional metrics we would like to collect and report during training. <br>

This case, we will use cross entropy as the loss argument. This loss is for a binary classification problem and is defined in Keras as "binary_crossentropy". <br>

We will define the optimizer as the efficient stochastic gradient algorithm "adam". This is a popular version of gradient descent because it automatically turns itself and gives good results in a wide range of problems. <br>

Finally, because it is a classification problem, we will collect and report the classfication accuracy, defined via the metrics argument.

In [32]:
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

## **4. Fit Keras Model**

We have defined our model and compiled it ready for efficient training. <br>

Now it is time to execute the model on some data <br>

Training occurs over **epochs** and each epoch is split into **batches**. <br>

1. **Epochs**: One pass through all of the row in the training data set. <br>
2. **Batches**: One or more samples considered by the model within an epoch before weights are updated. <br>

The training process will run for a fixed numberof iterations through the dataset called epochs, that we must specify using the epochs argument. We must also set the number of dataset rows that are considered before the model weights are updated within each epoch, called the batch size and set using the batch_size argument. <br>

For this problem, we will run for a small number of epochs (150) an use a relatively small batch size of 10. <br>

These configurations can be chosen experimentally by trial  and error. We want to train the model enough o it learns a good (or good enough) mapping of rows of input data to the output classification. The model will always have some error, but the emount of error will leve out after some point for a given model configuration. This is called model convergence.


In [33]:
model.fit(X,y,epochs=150,batch_size=10)

Epoch 1/150
77/77 [==============================] - 1s 1ms/step - loss: 7.6550 - accuracy: 0.4570
Epoch 2/150
77/77 [==============================] - 0s 1ms/step - loss: 1.8052 - accuracy: 0.5651
Epoch 3/150
77/77 [==============================] - 0s 1ms/step - loss: 1.4214 - accuracy: 0.5859
Epoch 4/150
77/77 [==============================] - 0s 2ms/step - loss: 1.2607 - accuracy: 0.6068
Epoch 5/150
77/77 [==============================] - 0s 2ms/step - loss: 1.1662 - accuracy: 0.6016
Epoch 6/150
77/77 [==============================] - 0s 2ms/step - loss: 1.1013 - accuracy: 0.6276
Epoch 7/150
77/77 [==============================] - 0s 2ms/step - loss: 1.0691 - accuracy: 0.5990
Epoch 8/150
77/77 [==============================] - 0s 1ms/step - loss: 0.9574 - accuracy: 0.6224
Epoch 9/150
77/77 [==============================] - 0s 1ms/step - loss: 0.9353 - accuracy: 0.6055
Epoch 10/150
77/77 [==============================] - 0s 1ms/step - loss: 0.9231 - accuracy: 0.6198
Epoch 11/

We can see that we have **768** records, so out of that, we choose **10** records at a time(referred to as the batch size) for **1** iteration (referred to as one Epoch).

## **5.Evaluate Keras Model**

We have trained our neural networks on the entire training dataset and we can evaluate the performance of the network on the same dataset. <br>

You an evaluate your model on your training dataset using the evaluate() function on your model and pass it the same input and output used to train the model. <br>

This will generate a prediction for each input and output pair and collect scores, including the average loss and any metrics you have configured, such as accuracy, <br>

The evaluate() function will return a list with two values. The first will be the lost of the model on the dataset and second will be the accuracy of the model on the dataset.

In [40]:
loss, accuracy = model.evaluate(X,y)
print('Accuracy: {}\nLoss: {}'.format(accuracy*100,loss))

24/24 [==============================] - 0s 1ms/step - loss: 0.5615 - accuracy: 0.7174
Accuracy: 71.74479365348816
Loss: 0.5615082383155823


## **7. Make predictions**

Making predictions is easy as calling the predict() function. We are using a sigmoid activation function on the output layer, so the predictions will be a probability in the range between 0 and 1. We can easily convert them into a crisp binary prediction for this classification task by rounding them.

In [38]:
predictions = model.predict(X)
rounded = [round(x[0]) for x in predictions]
rounded

24/24 [==============================] - 0s 650us/step


[0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,


In [39]:
predictions = (model.predict(X) > 0.5).astype(int)
#summarize the first 20 cases
for i in range(20):
    print('%s => %d (expected %d)' % (X[i].tolist(), predictions[i],y[i]))

24/24 [==============================] - 0s 650us/step
[6.0, 148.0, 72.0, 35.0, 0.0, 33.6, 0.627, 50.0] => 0 (expected 1)
[1.0, 85.0, 66.0, 29.0, 0.0, 26.6, 0.351, 31.0] => 0 (expected 0)
[8.0, 183.0, 64.0, 0.0, 0.0, 23.3, 0.672, 32.0] => 1 (expected 1)
[1.0, 89.0, 66.0, 23.0, 94.0, 28.1, 0.167, 21.0] => 0 (expected 0)
[0.0, 137.0, 40.0, 35.0, 168.0, 43.1, 2.288, 33.0] => 1 (expected 1)
[5.0, 116.0, 74.0, 0.0, 0.0, 25.6, 0.201, 30.0] => 0 (expected 0)
[3.0, 78.0, 50.0, 32.0, 88.0, 31.0, 0.248, 26.0] => 0 (expected 1)
[10.0, 115.0, 0.0, 0.0, 0.0, 35.3, 0.134, 29.0] => 0 (expected 0)
[2.0, 197.0, 70.0, 45.0, 543.0, 30.5, 0.158, 53.0] => 0 (expected 1)
[8.0, 125.0, 96.0, 0.0, 0.0, 0.0, 0.232, 54.0] => 0 (expected 1)
[4.0, 110.0, 92.0, 0.0, 0.0, 37.6, 0.191, 30.0] => 0 (expected 0)
[10.0, 168.0, 74.0, 0.0, 0.0, 38.0, 0.537, 34.0] => 1 (expected 1)
[10.0, 139.0, 80.0, 0.0, 0.0, 27.1, 1.441, 57.0] => 0 (expected 0)
[1.0, 189.0, 60.0, 23.0, 846.0, 30.1, 0.398, 59.0] => 1 (expected 1)
[5.0, 16